[pytorch] simplify tensor options logic in pybinding codegen

ljk53 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #46978 [pytorch] continue to rewrite gen_python_functions.py with typed models
* #46977 [pytorch] get rid of cpp_type_str from pybind codegen
* **#46976 [pytorch] simplify tensor options logic in pybinding codegen**
* #46243 [pytorch] add script to run all codegen

Technically, it's not semantic preserving, e.g.: emition of
'requires_grad' is no longer gated by 'has_tensor_return' - there is no
guarantee that is_like_or_new_function should all have tensor return.
But the output is identical so there might be some invariant - could
also add assertion to fail loudly when it's broken.

Differential Revision: [D24589211](https://our.internmc.facebook.com/intern/diff/D24589211)

ljk53 committed:
[pytorch] simplify tensor options logic in pybinding codegen

Technically, it's not semantic preserving, e.g.: emition of
'requires_grad' is no longer gated by 'has_tensor_return' - there is no
guarantee that is_like_or_new_function should all have tensor return.
But the output is identical so there might be some invariant - could
also add assertion to fail loudly when it's broken.

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] simplify tensor options logic in pybinding codegen"


Technically, it's not semantic preserving, e.g.: emition of
'requires_grad' is no longer gated by 'has_tensor_return' - there is no
guarantee that is_like_or_new_function should all have tensor return.
But the output is identical so there might be some invariant - could
also add assertion to fail loudly when it's broken.

[ghstack-poisoned]

ezyang reviewed (approved):
Shorter, and the new formulation makes sense.

ljk53 committed:
Update on "[pytorch] simplify tensor options logic in pybinding codegen"


Technically, it's not semantic preserving, e.g.: emition of
'requires_grad' is no longer gated by 'has_tensor_return' - there is no
guarantee that is_like_or_new_function should all have tensor return.
But the output is identical so there might be some invariant - could
also add assertion to fail loudly when it's broken.

Differential Revision: [D24589211](https://our.internmc.facebook.com/intern/diff/D24589211)

[ghstack-poisoned]
